### PR TITLE
Update Planckton to use HOOMD 3

### PR DESCRIPTION
**Pull request by gwenwhite**
This is a work in progress and is NOT ready to be merged. Just creating the pull request so discussions/questions can be added along the way.

### CODE REVIEW

File: planckton/sim.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def __init__(
     def run(self):
         """Run the simulation."""
         hoomd_args = f"--single-mpi --mode={self.mode}"
-        sim = hoomd.context.initialize(hoomd_args)
+        sim = hoomd.Simulation(device=hoomd_args)
```
Comment (chrisjonesBSU) [inline planckton/sim.py]: In hoomd3, the device object is its own class, so we might want to do something like:

device = hoomd.device.auto_select()
sim = hoomd.Simulation(device=device)

File: planckton/sim.py [hunk 2]
```diff
@@ -139,7 +139,7 @@ def __init__(
     def run(self):
         """Run the simulation."""
         hoomd_args = f"--single-mpi --mode={self.mode}"
```
Comment (chrisjonesBSU) [inline planckton/sim.py]: We can delete this line for hoomd3

**Comment by chrisjonesBSU** [comment]
Some things left to do:

1. Fix e-factor implementation
2. Update wall time check using hoomd3's wall time API
3. Update versions in environment files

**Review by chrisjonesBSU — APPROVED** [review]

File: environment.yml [hunk 3]
```diff
@@ -4,23 +4,23 @@ channels:
...
+  - foyer=0.11.1
   - git
-  - gmso=0.6.0
-  - gsd=2.4.2
-  - hoomd=2.9.6
-  - mbuild=0.13.0
-  - numpy=1.21
+  - gmso=0.8.1
+  - gsd=2.5.3
+  - hoomd=3.3.0
```
Comment (marjanalbooyeh) [inline environment.yml]: HOOMD just released version 3.4 last week. I checked the CHANGELOG and seems like nothing changed for the stuff we use here. I think we can safely use v3.4 in our environments now.

File: planckton/sim.py [hunk 4]
```diff
@@ -38,18 +34,19 @@ class Simulation:
...
     target_length : unyt.unyt_quantity, default None
         Target final box length for the shrink step. If None is provided, no
         shrink step will be performed.
     restart : str, default None
         Path to gsd file from which to restart the simulation.
-    nlist : str, default "cell"
-        Type of neighborlist to use. Options are "cell", "tree", and "stencil".
+    nlist : str, default "Cell"
+        Type of neighborlist to use. Options are "Cell", "Tree", and Stencil".
         See https://hoomd-blue.readthedocs.io/en/stable/nlist.html and
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: First link doesn't work. Probably because of the recent HOOMD update.

File: planckton/sim.py [hunk 5]
```diff
@@ -129,138 +129,134 @@ def __init__(
...
-                auto_scale=True,
-                restart=self.restart,
-                nlist=self.nlist,
-                r_cut=self.r_cut,
-            )
-            self.ref_values = ref_values
-            snap = hoomd_objects[0]
-            hoomd.util.unquiet_status()
+        # mbuild units are nm, amu
+        snap, hoomd_objects, ref_values = create_hoomd_forcefield(
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: I suggest renaming `hoomd_objects` to `hoomd_forcefield` for clarity. The second returned item is the list of hoomd forces.

File: planckton/sim.py [hunk 6]
```diff
@@ -129,138 +129,134 @@ def __init__(
...
+            self.system, auto_scale=True, r_cut=self.r_cut
+        )
+        if self.e_factor != 1:
+            lj = hoomd_objects[0]
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: Better to make sure this `lj` is actually an instance of the LJ force.
```
if isinstance(lj, hoomd.md.pair.LJ):
    for pair in lj.params:
        lj.params[pair]["epsilon"] *= self.e_factor
```

File: planckton/sim.py [hunk 7]
```diff
@@ -129,138 +129,134 @@ def __init__(
...
-                filename="trajectory.gsd",
-                period=self.gsd_write,
-                group=all_particles,
-                overwrite=False,
-                phase=0,
-                dynamic=["momentum"],
+        if self.target_length is not None:
+            # Run the shrink step
+            final_length = self.target_length.to("Angstrom").value
+            final_box = (self.shrink_steps, final_length)
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: `final_box` variable is overwritten in line 198.

File: planckton/sim.py [hunk 8]
```diff
@@ -129,138 +129,134 @@ def __init__(
...
-                integrator.randomize_velocities(seed=42)
-                hoomd.run_upto(self.shrink_steps)
-                box_resize.disable()
-                self.n_steps = [i + self.shrink_steps for i in self.n_steps]
+            sim.operations.updaters.append(box_resize)
+            sim.state.thermalize_particle_momenta(
+                filter=all_particles, kT=self.shrink_kT
+            )
+            sim.run(self.shrink_steps, write_at_start=True)
+            self.n_steps = [i + self.shrink_steps for i in self.n_steps]
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: Not sure if I understood the purpose of `n_steps` correctly but if `self.n_steps` is a list of number of steps for each simulation block, it means the values in that list indicate the length of each simulation block, not the final timestep at each block. In that case, you don't need to add the `shrink_steps` here.

File: planckton/sim.py [hunk 9]
```diff
@@ -129,138 +129,134 @@ def __init__(
...
 
-                try:
-                    hoomd.run_upto(n_steps + 1, limit_multiple=self.gsd_write)
-                    if sim.system.getCurrentTimeStep() >= self.n_steps[-1]:
-                        print("Simulation completed")
-                        done = True
-                except hoomd.WalltimeLimitReached:
-                    print("Walltime limit reached")
+                sim.run(n_steps)
+                if sim.timestep >= self.n_steps[-1]:
```
Comment (marjanalbooyeh) [inline planckton/sim.py]: Same comment here. 
For example, let's say `n_steps = [1000, 2000, 1000]`. At this line, `sim.timestep` is `4000 + shrink_steps`. Therefore, this if statement probably doesn't serve its purpose. 
You can check `sim.timestep > (sum of all items in n_step + shrink_steps`.

**Review by marjanalbooyeh — APPROVED** [review]
Looks very good to me. Thank you Gwen and Chris for the effort.
Just added a few minor comments.

**Pull request closed without merging**